### ISSUE TITLE
not handling disappearance of players correctly

### ISSUE THREAD

**Issue by cguthardt** [jishi/node-sonos-http-api#597]
I use node-sonos on a raspberry pi to control my Sonos system. Recently I started to use wireless plugs to semi-automatically power by sonos speakers on and off according to when I need them (as they suck a lot of energy for being idle 90% of the time).

Node-sonos seems to have issues with disappearing (due to power cut) players, it kind of seems to freeze in the last state. 

Example: I query the state of player in room "Kueche" while it's playing a song, everything works fine. I then remove power to that player, and query it again: node-sonos does not give an error, instead I get

`{"currentTrack":{"artist":"","title":"","album":"","albumArtUri":"","duration":0,"uri":"","type":"track"},"nextTrack":{"artist":"","title":"","album":"","albumArtUri":"","duration":0,"uri":""},"volume":12,"mute":false,"trackNo":1,"elapsedTime":11553,"elapsedTimeFormatted":"03:12:33","playbackState":"PLAYING","playMode":{"repeat":"none","shuffle":false,"crossfade":false}}`

The time elapsed keeps counting up, even though the song was just two minutes long. 

Next, I power up one of my other players that was formerly unplugged. It appears in the Sonos app and I can play songs on it, of course, but node-sonos doesn't recognize it. Instead, I get the following error messages on the log:

`2018-02-23T10:30:05.943Z ERROR 
  { error: 'action \'wohnzimmer\' not found' }
2018-02-23T10:30:48.807Z ERROR 
  TypeError: Cannot read property 'members' of undefined
    at Player.recalculateGroupVolume (/home/pi/node-sonos/node_modules/sonos-discovery/lib/prototypes/Player/recalculateGroupVolume.js:5:27)
    at NotificationListener.notificationHandler (/home/pi/node-sonos/node_modules/sonos-discovery/lib/models/Player.js:338:25)
    at emitTwo (events.js:92:20)
    at NotificationListener.emit (events.js:172:7)
    at /home/pi/node-sonos/node_modules/sonos-discovery/lib/NotificationListener.js:125:17
    at process._tickCallback (node.js:366:9)
2018-02-23T10:31:51.292Z ERROR 
  { error: 'action \'esszimmer\' not found' }`

After restarting node-sonos, everything works well again (until I try to query a powered-off speaker again).

This issue is really too bad, as with the Pi and node-sonos it would be amazing to use node-sonos to see if a speaker is actually power up, and if not run an IFTTT command to turn on the wifi plug that controls the speaker.

**Comment by cguthardt** [comment]
Another example:  

I run http://gutpi:5005/Kueche/zones and it shows me which players are currently connected: Kueche and Esszimmer.

Now I unplug Esszimmer and plug Wohnzimmer. This gets recognized by the Sonos app, of course, and Esszimmer diesappears and Wohnzimmer appears. So I run above command again: it still tells me that Kueche and Esszimmer are available! If I open [Kueche-IP]:1400/support/review  it shows that the player Kueche is well aware that Esszimmer is gone and now Wohnzimmer is there. It seems like node-sonos hangs in the previous state.

**Comment by jishi** [comment]
Which version are you running? The `/zones` should be up to date if the Sonos app actually doesn't see them anymore, since they should be based on the same topology events. 

However, the fact that you can query about a player that has been powered off I could see happening, because I might not care about removing players that is no longer in the zones info (although, I thought I did forget about them). The time elapsed is dynamically calculated based on the latest update form the player, to make the response to be as accurate as possible, hence the infinite ticking because I don't know that it has stopped. 

There might be some easy fixes for some of these issues, but from the beginning the assumption was made that players are always on (since that is the recommendation from Sonos) so it hasn't been a huge priority (or need) to support this.

**Comment by cguthardt** [comment]
Seems like I was running version 1.0, no idea where I had gotten that from, I only installed it in December. Now I upgraded to 1.6 and the behaviour is different. I will look into it more in the coming days and report back.

Btw, this recommendation from Sonos made some sense a few years ago, when Sonos would always run on its mesh network. Now with being able to run on normal wifi, it's really not necessary to keep all players on at all times, especially when most of the day I'm either at work or asleep (no, not at the same time ;-)
If they had gotton the power consumption on all of them down to the 2.2W of the new P5 it would be a different story.
